src/ucp/core/ucp_mm.c

```diff
@@ -1305,3 +1406,220 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+    ucs_status_t status;
+    ucp_memh_exported_unpacked_t unpacked;
+    ucs_rcache_region_t *rregion;
+    khiter_t iter;
+
+    status = ucp_memh_exported_unpack(context, export_mkey_buffer, &unpacked);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucp_memh_create(context, unpacked.address, unpacked.length,
+                             unpacked.mem_type, UCT_ALLOC_METHOD_LAST,
+                             UCP_MEM_FLAG_IMPORTED, &memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    if (ucs_likely(context->imported_mem_rcaches != NULL)) {
+        /* Try to find rcache of imported memory buffers for the specific peer with
+         * UUID packed in the exported_mkey_buffer */
+        UCP_THREAD_CS_ENTER(&context->mt_lock);
+
+        iter = kh_get(ucp_context_imported_mem_rcaches_hash,
+                      context->imported_mem_rcaches, unpacked.remote_uuid);
+        if (iter != kh_end(context->imported_mem_rcaches)) {
+            /* Found rcache for the specific peer */
+            rcache = kh_value(context->imported_mem_rcaches, iter);
+            ucs_assert(rcache != NULL);
+
+            rregion = ucs_rcache_lookup_unsafe(rcache, unpacked.address,
+                                               unpacked.length,
+                                               PROT_READ | PROT_WRITE);
+            if (rregion != NULL) {
+                rcache_memh = ucs_derived_of(rregion, ucp_mem_t);
+                if (ucs_likely(rcache_memh->reg_id == unpacked.reg_id)) {
+                    ucp_memh_rcache_print(rcache_memh, unpacked.address,
+                                          unpacked.length);
+                    memh->parent = rcache_memh;
+                    status       = UCS_OK;
+                    UCP_THREAD_CS_EXIT(&context->mt_lock);
+                    goto out_memh_update;
+                }
+
+                /* If registration IDs are not matched, but a region still
+                 * exists in the RCACHE of imported regions, it means that
+                 * an exported memory handle has already been destroyed for a
+                 * given address, but an imported memory handle hasn't been
+                 * retrieved from the RCACHE. So, it should have reference
+                 * counter == 1, since ucp_mem_unmap() should be invoked for
+                 * imported memory handles and then - for exported ones */
+                ucs_assertv(rregion->refcount == 1, "%u", rregion->refcount);
+                ucs_rcache_region_invalidate(rcache, rregion,
+                                             ucs_empty_function, NULL);
+            }
+        }
+
+        UCP_THREAD_CS_EXIT(&context->mt_lock);
+    }
+
+    status = ucp_memh_import_slow(context, rcache, memh, &unpacked);
+    if (status != UCS_OK) {
+        goto err_memh_free;
+    }
+
+out_memh_update:
+    if (memh->parent != memh) {
+        ucp_memh_init_from_parent(memh, memh->parent->md_map);
```

Comment:
seems you can move it to lint 1585 and remove this if

Reply:
unfortunately, we can't:
if `ucp_memh_import_slow` was successful, we should fill memh from parent's memh too